### ISSUE TITLE
Other types of Lotus::Model::Error for some specific adapter exceptions

### ISSUE THREAD

**Issue by hlegius** [hanami/model#284]
Trying out Lotus Model v0.5.2 here, I've found that Lotus::Model::Error missed a more specialized way to deal with Sequel Exceptions.

For example: I need to deal with `Sequel::UniqueConstraintViolation` for any repository I might have. My PK’s are Postgres’ _uuid fields_ which throws that exception if I try to query any data but UUID Format. Something like this:

```ruby
my_non_uuid = 1234
MyRepository.find(my_non_uuid) # fail with Sequel::UniqueCo…
MyRepository.find('82994efa-5d21-4e8a-9a25-0c9448b91f4e') # nil or Object :+1:
```

In 0.5.2, I will get a generic `Lotus::Model::Error` which make a bit harder to identify this kind of contract violation of my requester. I am not suggesting to Lotus deals with _uniqueness constraints_. Just a (IMO) _better_ encapsulation over application layer/tiers.

TL;DR: my suggestion downs to: more (and specialized) Lotus Model exceptions.

**Comment by joneslee85** [comment]
@hlegius the implementation allows you to query the original exception raised via `Exception#cause`:

```
begin
  my_non_uuid = 1234
  MyRepository.find(my_non_uuid)
rescue => error
  puts "Current failure: #{error.inspect}" #fail with Lotus::Model::Error
  puts "Original failure:  #{error.cause.inspect}" #fail with Sequel::UniqueCo…
end
```

it is a bit awkward but works for now

**Comment by joneslee85** [comment]
cc @jodosha

**Comment by hlegius** [comment]
I see, thanks for quick replay! Yep, it works, but we'll need to always `rescue` and use `error.cause`. I've got into a case when I'd declare what Exception to expect, i.e: 

```ruby
# Grape API
rescue_from MyWeirdoException do |exception|
 # ... do something
end
```

Maybe it worth to add more Lotus Exceptions someday later, IMO.

**Comment by jodosha** [comment]
@hlegius This is a good catch. I made a mistake by suggesting to @Moratorius to unify all the errors raised by Sequel into a generic `Lotus::Model::Error`. See https://github.com/hanami/model/pull/281

This doesn't allow to rescue from specific and precious Sequel exceptions like unique constraint violations. With this change there is no way to handle record uniqueness. My apologize to @Moratorius Who suggested the right thing to do.

@hlegius @Moratorius @joneslee85 Anyone wants to restore the initial implementation of #281 by @Moratorius ? We should have our own exceptions for these cases.

Thank you all and sorry again about this mistake.

**Comment by hlegius** [comment]
I can take care of it, @jodosha o/
